File: 2025/04 April/15 Bellman-Ford/main.py
# Link: https://www.geeksforgeeks.org/problems/distance-from-the-source-bellman-ford-algorithm/1


class Solution:
    def bellmanFord(self, V, edges, src):
        INF = 10**8
        dist = [INF for _ in range(V)]
        dist[src] = 0

        for _ in range(V - 1):
            for s, e, dis in edges:
                if dist[s] != INF and dist[s] + dis < dist[e]:
                    dist[e] = dist[s] + dis

        for s, e, dis in edges:
            if dist[s] != INF and dist[s] + dis < dist[e]:
                return [-1]

        return dist


if __name__ == "__main__":
    obj = Solution()
    V = 5
    edges = [[1, 3, 2], [4, 3, -1], [2, 4, 1], [1, 2, 1], [0, 1, 5]]
    src = 0
    result = obj.bellmanFord(V, edges, src)
    print(result)

    V = 4
    edges = [[0, 1, 4], [1, 2, -6], [2, 3, 5], [3, 1, -2]]
    src = 0
    result = obj.bellmanFord(V, edges, src)
    print(result)

File: 2025/04 April/15 Bellman-Ford/main.cpp
// Link: https://www.geeksforgeeks.org/problems/distance-from-the-source-bellman-ford-algorithm/1

#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> bellmanFord(int V, vector<vector<int>> &edges, int src)
    {
        vector<int> dist(V, 1e8);
        dist[src] = 0;

        for (int i = 0; i < V; i++)
        {

            for (vector<int> edge : edges)
            {
                int u = edge[0];
                int v = edge[1];
                int wt = edge[2];
                if (dist[u] != 1e8 && dist[u] + wt < dist[v])
                {

                    if (i == V - 1)
                        return {-1};

                    dist[v] = dist[u] + wt;
                }
            }
        }

        return dist;
    }
};

int main()
{
    Solution obj;
    int V = 5;
    vector<vector<int>> edges = {{1, 3, 2}, {4, 3, -1}, {2, 4, 1}, {1, 2, 1}, {0, 1, 5}};
    int src = 0;
    vector<int> result = obj.bellmanFord(V, edges, src);
    for (int i : result)
    {
        cout << i << " ";
    }
    cout << endl;

    V = 4;
    edges = {{0, 1, 4}, {1, 2, -6}, {2, 3, 5}, {3, 1, -2}};
    src = 0;
    result = obj.bellmanFord(V, edges, src);
    for (int i : result)
    {
        cout << i << " ";
    }
    cout << endl;
    return 0;
}